src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -1,31 +1,242 @@
+                return false;
+            }
+        } else {
+            log.warn("IN CategoryServiceImpl (removeCategoryById): the category with id {} has not been deleted",
+                    categoryId);
+            return false;
+        }
+    }
+
+    @Override
+    public boolean removeSubcategoryById(Long subcategoryId) {
+        if (isSubcategoryDeletable(subcategoryId)) {
+            subcategoryRepository.deleteById(subcategoryId);
+            log.info("IN CategoryServiceImpl (removeSubcategoryById): the subcategory with id {} has been deleted successfully",
+                    subcategoryId);
+            return true;
+        } else {
+            log.warn("IN CategoryServiceImpl (removeSubcategoryById): the subcategory with id {} has not been deleted",
+                    subcategoryId);
+            return false;
+        }
+    }
+
+    private boolean isCategoryDtoIdValid(Long id) {
+        if (!categoryRepository.existsById(id)) {
+            log.warn("IN CategoryServiceImpl (isCategoryDtoIdInvalid): the category does not exist by this id: {}", id);
+            return false;
+        }
+        return !isIdInvalid(id);
+    }
+
+    private boolean isIdInvalid(Long id) {
+        if (id == null || id <= 0L) {
+            log.warn("IN CategoryServiceImpl (isIdInvalid): this id is invalid {}", id);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isSubcategoryDtoIdInvalid(List<SubcategoryControllerDto> subcategoriesDtos) {
+        return subcategoriesDtos.stream()
+                .filter(subcategoryDto -> subcategoryDto.getId() != null)
+                .map(SubcategoryControllerDto::getId)
+                .anyMatch(this::isSubcategoryDtoIdInvalid);
+    }
+
+    private boolean isSubcategoryDtoIdInvalid(Long subcategoryId) {
+        if (isIdInvalid(subcategoryId) || !subcategoryRepository.existsById(subcategoryId)) {
+            log.warn("IN CategoryServiceImpl (isSubcategoryDtoIdInvalid): the subcategory does not exist by given id {}",
+                    subcategoryId);
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isSubcategoryDeletable(Long subcategoryDtoId) {
+        if (!isSubcategoryDtoIdInvalid(subcategoryDtoId)) {
+            Optional<Subcategory> subcategory = subcategoryRepository.findById(subcategoryDtoId);
+            if (subcategory.isPresent()) {
+                if (subcategory.get().getProducts().isEmpty()) {
+                    return true;
+                } else {
+                    log.warn("IN CategoryServiceImpl (isSubcategoryDeletable): the subcategory with id {} can not be deleted" +
+                                    " because it has products",
+                            subcategoryDtoId);
+                    return false;
+                }
+            } else {
+                log.warn("IN CategoryServiceImpl (isSubcategoryDeletable): the subcategory does not exist by given id {}",
```

Comment:
Please make a new method for logging that will get message and object rather copy/paste it all the time.